decorators.py

```diff
@@ -0,0 +1,164 @@
+
+        if not library:
+            return LIBRARY_NOT_FOUND.response
+
+        g.library = library
+
+        return f(*args, **kwargs)
+
+    return decorated
+
+
+def returns_problem_detail(f):
+    """
+    Allows a view function, on error, to return a specific ProblemDetail instance,
+    which will be rendered into a flask Response object.
+    """
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        v = f(*args, **kwargs)
+
+        if isinstance(v, ProblemDetail):
+            return v.response
+
+        return v
+
+    return decorated
+
+
+def returns_json_or_response_or_problem_detail(f):
+    """
+    Provides a usable Response for view functions which return any of:
+        - A Python data structure (will be run through flask.jsonify())
+        - A ProblemDetail instance
+        - An instance of flask.Response or a subclass thereof
+    """
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        v = f(*args, **kwargs)
+
+        if isinstance(v, ProblemDetail):
+            return v.response
+
+        if isinstance(v, Response):
+            return v
+
+        return jsonify(**v)
+
+    return decorated
+
+
+def compressible(f):
+    """
+    Compress the outgoing response payload using gzip if:
+        * The response being rendered by the view is not a ProblemDetail, etc.
+        * The request included an 'Accept-Encoding' header whose value includes 'gzip'
+        * The response being rendered by the view carries a 2xx status code
+        * The response body is not already explicitly encoded
+    """
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        response = f(*args, **kwargs)
+        accept_encoding = request.headers.get('Accept-Encoding', '')
+
+        if (                                                # Reasons to exit early, without compressing:
+            not isinstance(response, Response) or           # - We don't want to compress non-Response data
+            'gzip' not in accept_encoding.lower() or        # - They didn't ask for compression
+            not (199 < response.status_code < 300) or       # - It's not a 2xx response--we only compress 2xx
+            'Content-Encoding' in response.headers          # - It's already been encoded, don't mess with it
+        ):
+            return response
+
+        # Perform the compression on the response data
+        buffer = BytesIO()
+        with gzip.GzipFile(mode='wb', fileobj=buffer) as gzipped:
+            gzipped.write(response.data)
+        response.data = buffer.getvalue()
+
+        response.direct_passthrough = False  # This skips some Werkzeug/Flask checks, unneeded for a binary payload.
+        response.headers['Content-Encoding'] = 'gzip'
+        response.headers['Vary'] = 'Accept-Encoding'    # TODO: This is bad if Vary is already set.
+        response.headers['Content-Length'] = len(response.data)
+
+        return response
+
+    return decorated
```

Comment:
I'm getting the impression that you've moved the app_helpers into this file and will be removing or greatly cutting app_helpers.

Reply:
Right--I'll remove them from app_helpers once I confirm that nothing uses the older versions.